[Motions 2022 11 lwg 13] P2505R5 Monadic Functions for std::expected (#5993)
cplusplus/draft: source/utilities.tex

contains an object of type \tcode{T}.

\pnum

> A type \tcode{T} is a \term{valid value type for \tcode{expected}}, if

Reviewer: N.B. as per https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#latex-source-code-formatting we try to use "semantic line breaks" to minimise the size of diffs when we make later edits.

In this line, I would have put the line break after the comma, with the "if" on the next line. And in the line below, break at the end of the sentence, before "A program which ...".

There are lots of places in this PR that don't use semantic line breaks. I don't think that should block the merge, but please bear it in mind for future edits, thanks!

Reviewer 2: Looks like this has been addressed. Many thanks, @RobertLeahy!

\tcode{remove_cv_t<T>} is \tcode{void} or a complete non-array object type
that is not \tcode{in_place_t}, \tcode{unexpect_t}, or a specialization of

> \tcode{unexpected}. A program which instantiates class template

Reviewer 3: ```suggestion
\tcode{unexpected}. A program that instantiates the class template
```
```cli
$ grep 'program which' *.tex | wc -l
0
$ grep 'a program that' *.tex | wc -l
14
```

Reviewer 2: I agree with Walter that we shouldn't make natural language decisions purely based on ngram statistics, devoid of context or rationale.

More importantly, the paper is explicitly changing prior "program that" wording to "program which", so I would hesitate to immediately overrule the wording group's explicit output here. Generally, we apply the approved wording as is, and it's mainly presentational style details (like OK/error comments, or the equivalent-to wording you pointed out elsewhere) that we adjust. For the change you propose here I would much prefer a separate follow-up issue.

Reviewer 3: That makes sense. I should be more careful. Dismiss this. Since the current wording makes sense anyways, I'm not going to attempt pursuing this for consistency.

Reviewer 3: Opened https://github.com/cplusplus/draft/issues/5994 for this.

\tcode{has_value() ? std::move(**this) : static_cast<T>(std::forward<U>(v))}.
\end{itemdescr}

\indexlibrarymember{error_or}{expected}%
\begin{itemdecl}

> template <class G = E> constexpr E error_or(G&& e) const &;

Reviewer 3: Replace all `template <` with `template<`.

template<class F> constexpr auto and_then(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{U} be
\tcode{remove_cvref_t<invoke_result_t<F, decltype(std::move(value()))>>}.

\pnum
\constraints
\tcode{is_move_constructible_v<E>} is \tcode{true}.

\pnum
\mandates
\tcode{U} is a specialization of \tcode{expected} and
\tcode{is_same_v<U::error_type, E>} is \tcode{true}.

\pnum
\effects
Equivalent to:
\begin{codeblock}
if (has_value())
  return invoke(std::forward<F>(f), std::move(value()));
else
  return U(unexpect, std::move(error()));
\end{codeblock}
\end{itemdescr}

\indexlibrarymember{or_else}{expected}%
\begin{itemdecl}
template<class F> constexpr auto or_else(F&& f) &;
template<class F> constexpr auto or_else(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be \tcode{remove_cvref_t<invoke_result_t<F, decltype(error())>>}.

\pnum
\constraints
\tcode{is_copy_constructible_v<T>} is \tcode{true}.

\pnum
\mandates
\tcode{G} is a specialization of \tcode{expected} and
\tcode{is_same_v<G::value_type, T>} is \tcode{true}.

\pnum
\effects
Equivalent to:
\begin{codeblock}
if (has_value())
  return G(in_place, value());
else

>   return invoke(std::forward<F>(), error());

Reviewer 3: ```suggestion
  return invoke(std::forward<F>(f), error());
```

\end{codeblock}
\end{itemdescr}

\indexlibrarymember{or_else}{expected}%
\begin{itemdecl}
template<class F> constexpr auto or_else(F&& f) &&;
template<class F> constexpr auto or_else(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be
\tcode{remove_cvref_t<invoke_result_t<F, decltype(std::move(error()))>>}.

\pnum
\constraints
\tcode{is_move_constructible_v<T>} is \tcode{true}.

\pnum
\mandates
\tcode{G} is a specialization of \tcode{expected} and
\tcode{is_same_v<G::value_type, T>} is \tcode{true}.

\pnum
\effects
Equivalent to:
\begin{codeblock}
if (has_value())
  return G(in_place, std::move(value()));
else
  return invoke(std::forward<F>(f), std::move(error()));
\end{codeblock}
\end{itemdescr}

\indexlibrarymember{transform}{expected}%
\begin{itemdecl}
template<class F> constexpr auto transform(F&& f) &;
template<class F> constexpr auto transform(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{U} be

> \tcode{remove_cv_t<invoke_result_t<F, decltype(std::move(value()))>>}.

Reviewer: ```suggestion
\tcode{remove_cv_t<invoke_result_t<F, decltype(value())>>}.
```

\pnum
\constraints
\tcode{is_copy_constructible_v<E>} is \tcode{true}.

\pnum
\mandates

> U is a valid type for \tcode{expected}. If \tcode{is_void_v<U>} is

Reviewer: ```suggestion
\tcode{U} is a valid value type for \tcode{expected}. If \tcode{is_void_v<U>} is
```

Reviewer: Looks like the `U` got fixed in the new push, but the "value" is still missing.

Reviewer: Fixed in bb80c20a5230cbd3ab48050a4141b3cfb8f648f8

U u(invoke(std::forward<F>(f), value()));
\end{codeblock}
is well-formed.

\pnum
\effects
\begin{itemize}
\item
If \tcode{has_value()} is \tcode{false}, returns
\tcode{expected<U, E>(unexpect, error())}.
\item
Otherwise, if \tcode{is_void_v<U>} is \tcode{false}, returns an
\tcode{expected<U, E>} object whose \exposid{has_val} member is \tcode{true}
and \exposid{val} member is direct-non-list-initialized with
\tcode{invoke(std::forward<F>(f), value())}.
\item
Otherwise, evaluates \tcode{invoke(std::forward<F>(f), value())} and then
returns \tcode{expected<U, E>()}.
\end{itemize}
\end{itemdescr}

\indexlibrarymember{transform}{expected}%
\begin{itemdecl}
template<class F> constexpr auto transform(F&& f) &&;
template<class F> constexpr auto transform(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{U} be
\tcode{remove_cv_t<invoke_result_t<F, decltype(std::move(value()))>>}.

\pnum
\constraints
\tcode{is_move_constructible_v<E>} is \tcode{true}.

\pnum
\mandates
\tcode{U} is a valid value type for \tcode{expected}. If \tcode{is_void_v<U>} is
\tcode{false}, the declaration
\begin{codeblock}
U u(invoke(std::forward<F>(f), std::move(value())));
\end{codeblock}

> is well-formed.

Reviewer: In the similar wording for `optional` we say "is well-formed for some invented variable `u`" here. I don't think it matters.

Reviewer 2: Good idea, done!

\pnum
\effects
\begin{itemize}
\item
If \tcode{has_value()} is \tcode{false}, returns
\tcode{expected<U, E>(unexpect, std::move(error()))}.
\item
Otherwise, if \tcode{is_void_v<U>} is \tcode{false}, returns an
\tcode{expected<U, E>} object whose \exposid{has_val} member is \tcode{true}
and \exposid{val} member is direct-non-list-initialized with
\tcode{invoke(std::forward<F>(f), std::move(value()))}.
\item
Otherwise, evaluates \tcode{invoke(std::forward<F>(f), std::move(value()))} and
then returns \tcode{expected<U, E>()}.
\end{itemize}
\end{itemdescr}

\indexlibrarymember{transform_error}{expected}%
\begin{itemdecl}
template <class F> constexpr auto transform_error(F&& f) &;
template <class F> constexpr auto transform_error(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be \tcode{remove_cv_t<invoke_result_t<F, decltype(error())>>}.

\pnum
\constraints
\tcode{is_copy_constructible_v<T>} is \tcode{true}.

\pnum
\mandates
\tcode{G} is a valid value type for \tcode{expected} and the declaration
\begin{codeblock}
G g(invoke(std::forward<F>(f), error()));
\end{codeblock}
is well-formed.

\pnum
\returns
If \tcode{has_value()} is \tcode{true},

> \tcode{expected<T, G>(in_place, value())}; otherwise an \tcode{expected<T, G>}

Reviewer 3: ```suggestion
\tcode{expected<T, G>(in_place, value())}; otherwise, an \tcode{expected<T, G>}
```

\pnum
\mandates
\tcode{is_copy_constructible_v<E>} is \tcode{true} and
\tcode{is_convertible_v<G, E>} is \tcode{true}.

\pnum
\returns
\tcode{std::forward<G>(e)} if \tcode{has_value()} is \tcode{true},
\tcode{error()} otherwise.
\end{itemdescr}

\indexlibrarymember{error_or}{expected<void>}%
\begin{itemdecl}
template <class G = E> constexpr E error_or(G&& e) &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
\mandates
\tcode{is_move_constructible_v<E>} is \tcode{true} and
\tcode{is_convertible_v<G, E>} is \tcode{true}.

\pnum
\returns
\tcode{std::forward<G>(e)} if \tcode{has_value()} is \tcode{true},
\tcode{std::move(error())} otherwise.
\end{itemdescr}

\rSec3[expected.void.monadic]{Monadic operations}

\indexlibrarymember{and_then}{expected<void>}%
\begin{itemdecl}
template <class F> constexpr auto and_then(F&& f) &;
template <class F> constexpr auto and_then(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{U} be \tcode{remove_cvref_t<invoke_result_t<F>>}.

\pnum
\constraints
\tcode{is_copy_constructible_v<E>} is \tcode{true}.

\pnum
\mandates
\tcode{U} is a specialization of \tcode{expected} and
\tcode{is_same_v<U::error_type, E>} is \tcode{true}.

\pnum
\effects

> Equivalent to

Reviewer 3: ```suggestion
Equivalent to:
```

\begin{codeblock}
if (has_value())
  return invoke(std::forward<F>(f));
else
  return U(unexpect, error());
\end{codeblock}
\end{itemdescr}

\indexlibrarymember{and_then}{expected<void>}%
\begin{itemdecl}
template <class F> constexpr auto and_then(F&& f) &&;
template <class F> constexpr auto and_then(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{U} be \tcode{remove_cvref_t<invoke_result_t<F>>}.

\pnum
\constraints
\tcode{is_move_constructible_v<E>} is \tcode{true}.

\pnum
\mandates
\tcode{U} is a specialization of \tcode{expected} and
\tcode{is_same_v<U::error_type, E>} is \tcode{true}.

\pnum
\effects

> Equivalent to

Reviewer 3: ```suggestion
Equivalent to:
```

\begin{codeblock}
if (has_value())
  return invoke(std::forward<F>(f));
else
  return U(unexpect, std::move(error()));
\end{codeblock}
\end{itemdescr}

> \indexlibrarymember{and_then}{expected<void>}%

Reviewer 3: ```suggestion
\indexlibrarymember{or_else}{expected<void>}%
```

\begin{itemdecl}
template <class F> constexpr auto or_else(F&& f) &;
template <class F> constexpr auto or_else(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be \tcode{remove_cvref_t<invoke_result_t<F, decltype(error())>>}.

\pnum
\mandates
\tcode{G} is a specialization of \tcode{expected} and
\tcode{is_same_v<G::value_type, T>} is \tcode{true}.

\pnum
\effects

> Equivalent to

Reviewer 3: ```suggestion
Equivalent to:
```

\begin{codeblock}
if (has_value())
  return G();
else

>   return invoke(std::forward<F>(), error());

Reviewer 3: ```suggestion
  return invoke(std::forward<F>(f), error());
```

\end{codeblock}
\end{itemdescr}

> \indexlibrarymember{and_then}{expected<void>}%

Reviewer 3: ```suggestion
\indexlibrarymember{or_else}{expected<void>}%
```

\begin{itemdecl}
template <class F> constexpr auto or_else(F&& f) &&;
template <class F> constexpr auto or_else(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be
\tcode{remove_cvref_t<invoke_result_t<F, decltype(std::move(error()))>>}.

\pnum
\mandates
\tcode{G} is a specialization of \tcode{expected} and
\tcode{is_same_v<G::value_type, T>} is \tcode{true}.

\pnum
\effects

> Equivalent to

Reviewer 3: ```suggestion
Equivalent to:
```

template <class F> constexpr auto transform_error(F&& f) const &;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be \tcode{remove_cv_t<invoke_result_t<F, decltype(error())>>}.

\pnum
\mandates
\tcode{G} is a valid value type for \tcode{expected} and the declaration
\begin{codeblock}
G g(invoke(std::forward<F>(f), error()));
\end{codeblock}
is well-formed.

\pnum
\returns
If \tcode{has_value()} is \tcode{true}, \tcode{expected<T, G>()}; otherwise, an
\tcode{expected<T, G>} object whose \exposid{has_val} member is \tcode{false}
and \exposid{unex} member is direct-non-list-initialized with
\tcode{invoke(std::forward<F>(f), error())}.
\end{itemdescr}

\indexlibrarymember{transform_error}{expected<void>}%
\begin{itemdecl}
template <class F> constexpr auto transform_error(F&& f) &&;
template <class F> constexpr auto transform_error(F&& f) const &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
Let \tcode{G} be
\tcode{remove_cv_t<invoke_result_t<F, decltype(std::move(error()))>>}.

\pnum
\mandates
\tcode{G} is a valid value type for \tcode{expected} and the declaration
\begin{codeblock}
G g(invoke(std::forward<F>(f), std::move(error())));
\end{codeblock}
is well-formed.

\pnum
\returns

> If \tcode{has_value()} is \tcode{true}, \tcode{expected<T, G>()}; otherwise an

Reviewer 3: ```suggestion
If \tcode{has_value()} is \tcode{true}, \tcode{expected<T, G>()}; otherwise, an
```
